### PR TITLE
Hookup APD frontend to backend

### PR DESCRIPTION
**Pull request by mgwalker**
https://hitech-apd-pr544.app.cloud.gov/

- makes the main page "protected" (meaning, you have to login)
- on successful login or auth check, go ahead and fetch the user's APDs and use the first one returned as the current APD
  - an "auth check" happens when you first load the page - if you're already logged in, it just proceeds; otherwise it drops you at the login screen
  - there's some future work indicated here: how to pick which APD the user should be looking at, maybe there's a landing page before this, etc.
- populates the application state with the data from the API
- adds a temporary save button that saves the application state to the API

Addresses #394

### This pull request changes...
- can save and resume editing an APD

### This pull request is ready to merge when...
- [x] Tests have been updated (and all tests are passing)
- [ ] This code has been reviewed by someone other than the original author
- ~The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented~
- ~The change has been documented~
  - ~Associated OpenAPI documentation has been updated~

### This feature is done when...
- ~Design has approved the experience~
- ~Product has approved the experience~

### CODE REVIEW

File: web/src/routes.js [hunk 1]
```diff
@@ -6,10 +6,10 @@ import PoC from './components/PoC';
 import Login from './containers/Login';
 
 const routes = [
-  { path: '/', component: PoC, exact: true, nonPrivate: true },
+  { path: '/', component: PoC, exact: true, nonPrivate: false },
```
Comment (mgwalker) [inline web/src/routes.js]: You have to be authenticated in order to send the API requests added in this PR, so...  there's this.

File: web/src/actions/apd.js [hunk 2]
```diff
@@ -38,3 +48,100 @@ export const fetchApdDataIfNeeded = () => (dispatch, getState) => {
...
+      description: activity.descLong,
+      alternatives: activity.altApproach,
+      costAllocationMethodology: activity.costAllocateDesc,
+      otherFundingSources: {
+        description: activity.otherFundingDesc,
+        amount: activity.otherFundingAmt
+      },
+      goals: activity.goals.map(g => ({
+        description: g.desc,
+        objectives: [{ description: g.obj }] // TODO - objective should mape 1:1 to goals, instead of being an array (needs backend change first)
```
Comment (mgwalker) [inline web/src/actions/apd.js]: See #550. Once we figure out how the API will send/receive goals and objectives, we should update this.

File: web/src/actions/apd.test.js [hunk 3]
```diff
@@ -35,7 +35,7 @@ describe('state actions', () => {
 
     it('creates GET_APD_SUCCESS after successful APD fetch', () => {
       const store = mockStore({});
-      fetchMock.onGet('/apds').reply(200, { foo: 'bar' });
+      fetchMock.onGet('/apds').reply(200, [{ foo: 'bar' }]);
```
Comment (mgwalker) [inline web/src/actions/apd.test.js]: The `GET /apds` endpoint returns an array

File: web/src/components/Inputs.js [hunk 4]
```diff
@@ -9,10 +9,21 @@ class RichText extends Component {
...
+      lastContent: props.content
     };
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    if (nextProps.content !== this.state.lastContent) {
+      this.setState({
+        content: htmlToEditor(nextProps.content),
+        lastContent: nextProps.content
+      });
```
Comment (mgwalker) [inline web/src/components/Inputs.js]: Don't automatically return true here.  Doing so causes a non-stop render loop.  Instead, trigger a state update and wait for the state change before updating the component.

File: web/src/reducers/activities.js [hunk 5]
```diff
@@ -256,6 +258,96 @@ const reducer = (state = initialState, action) => {
...
+      const byId = {};
+      action.data.activities.forEach(a => {
+        byId[a.id] = {
+          id: a.id,
+          name: a.name,
+          types: ['HIT'], // TODO
+          descShort: '', // TODO
+          descLong: a.description,
+          altApproach: '', // TODO
+          costAllocateDesc: '', // TODO
```
Comment (mgwalker) [inline web/src/reducers/activities.js]: These 4 `TODO`s are partially represented in #554. After that one is done, we'll need to come back and update these.

File: web/src/reducers/activities.js [hunk 6]
```diff
@@ -256,6 +258,96 @@ const reducer = (state = initialState, action) => {
...
+          otherFundingDesc: a.otherFundingSources.description,
+          otherFundingAmt: +a.otherFundingSources.amount,
+          goals: a.goals.map(g => ({
+            desc: g.description,
+            obj: g.objectives[0] // TODO - don't assume there's one objective; tie objective directly to the goal instead of a mapped table
```
Comment (mgwalker) [inline web/src/reducers/activities.js]: #550

File: web/src/reducers/apd.js [hunk 7]
```diff
@@ -44,7 +44,15 @@ const reducer = (state = initialState, action) => {
         ...state,
         fetching: false,
         loaded: true,
-        data: { ...action.data[0] }
+        data: {
+          id: action.data.id,
+          years: ['2018'], // TODO
```
Comment (mgwalker) [inline web/src/reducers/apd.js]: #462 probably; lots to digest on this one

**Comment by mgwalker** [comment]
@brendansudol Sorry I added you as a reviewer earlier, but feeling pretty good this is ready to go now.

@NAretakis @jeromeleecms Note that the API enforces certain pieces of data integrity right now, and we don't have UI to indicate when the API rejects your save.  The biggest ones you're likely to run into are 1) activity names must not be empty and they must be unique within the APD; and 2) goals must not be empty (this one will be addressed in #550).

File: web/src/reducers/activities.js (hunk 5 above)
Comment (brendansudol) [reply web/src/reducers/activities.js]: 👍

**Pull request merged**